src/core/service.c

```diff
@@ -2705,99 +2708,175 @@ static void service_enter_reload_by_notify(Service *s) {
+        service_enter_running(s, SERVICE_SUCCESS);
+}
+
 static bool service_should_reload_extensions(Service *s) {
         int r;
 
         assert(s);
 
-        // TODO: Add support for user services, which can use ExtensionDirectories= + notify-reload.
-        // For now, skip for user services.
-        if (UNIT(s)->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
-                log_once(LOG_WARNING, "Not reloading extensions for user services.");
+        if (!pidref_is_set(&s->main_pid)) {
+                log_unit_debug(UNIT(s), "Not reloading extensions for service without main PID.");
                 return false;
         }
 
         r = exec_context_has_vpicked_extensions(&s->exec_context);
-        if (r < 0) {
+        if (r < 0)
                 log_unit_warning_errno(UNIT(s), r, "Failed to determine if service should reload extensions, assuming false: %m");
+        if (r <= 0)
+                return false;
+
+        // TODO: Add support for user services, which can use ExtensionDirectories= + notify-reload.
+        // For now, skip for user services.
+        if (UNIT(s)->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_once(LOG_WARNING, "Not reloading extensions for user services.");
                 return false;
         }
-        return r > 0;
+
+        return true;
 }
 
-static int service_reload_extensions(Service *s) {
-        // TODO: do this asynchronously
-        _cleanup_free_ char *propagate_dir = NULL;
-        PidRef *unit_pid = NULL;
+static void service_enter_reload_extensions(Service *s) {
+        _cleanup_(pidref_done) PidRef worker = PIDREF_NULL;
+        int r;
 
         assert(s);
 
-        unit_pid = unit_main_pid(UNIT(s));
-        if (!pidref_is_set(unit_pid)) {
-                log_unit_debug(UNIT(s), "Not reloading extensions for service without main PID.");
-                return 0;
+        /* If we don't have extensions to reload, immediately go to the signal step */
+        if (!service_should_reload_extensions(s))
+                return (void) service_enter_reload_signal(s);
+
+        service_unwatch_control_pid(s);
+        s->reload_result = SERVICE_SUCCESS;
+        s->control_command = NULL;
+        s->control_command_id = _SERVICE_EXEC_COMMAND_INVALID;
+
+        /* Given we are running from PID1, avoid doing potentially heavy I/O operations like opening images
+         * directly, and instead fork a worker process. */
+        r = unit_fork_helper_process(UNIT(s), "(sd-refresh-extensions)", /* into_cgroup= */ false, &worker);
+        if (r < 0) {
+                log_unit_error_errno(UNIT(s), r, "Failed to fork process to refresh extensions in unit's namespace: %m");
+                goto fail;
         }
+        if (r == 0) {
+                PidRef *unit_pid = unit_main_pid(UNIT(s));
+                if (!pidref_is_set(unit_pid)) {
+                        /* Mostly paranoia, as it was checked by the parent */
```

Comment:
it was? where?

Reply:
this really should be checked in the parent, and then just be an assert() here or so

Reply:
checked by service_should_reload_extensions() just before the fork